### PR TITLE
Implement "for" statements (and some supporting features) in LibJS

### PR DESCRIPTION
**Pull request by deoxxa**

### CODE REVIEW

File: Libraries/LibJS/AST.h [hunk 1]
```diff
@@ -211,13 +211,44 @@ class WhileStatement : public Statement {
...
+    virtual Value execute(Interpreter&) const override;
+    virtual void dump(int indent) const override;
+
+private:
+    virtual const char* class_name() const override { return "ForStatement"; }
+
+    OwnPtr<Statement> m_init;
+    OwnPtr<Expression> m_test;
+    OwnPtr<Expression> m_update;
+    NonnullOwnPtr<ScopeNode> m_body;
```
Comment (sunverwerth) [inline Libraries/LibJS/AST.h]: Body can be any kind of `Statement` it can even be empty

File: Libraries/LibJS/Parser.cpp [hunk 2]
```diff
@@ -215,6 +253,53 @@ NonnullOwnPtr<VariableDeclaration> Parser::parse_variable_declaration()
...
+NonnullOwnPtr<ForStatement> Parser::parse_for_statement()
+{
+    consume(TokenType::For);
+
+    consume(TokenType::ParenOpen);
+
+    OwnPtr<Statement> init = nullptr;
+    switch (m_current_token.type()) {
+    case TokenType::Var:
+        init = parse_variable_declaration();
```
Comment (sunverwerth) [inline Libraries/LibJS/Parser.cpp]: I think `parse_statement` is enough, as `VariableDeclaration` is a statement as well.

**Comment by awesomekling** [comment]
Very tasteful!

File: Libraries/LibJS/AST.h [hunk 3]
```diff
@@ -387,6 +418,10 @@ class CallExpression : public Expression {
 
 enum class AssignmentOp {
     Assign,
+    PlusEquals,
+    MinusEquals,
+    AsteriskEquals,
```
Comment (sunverwerth) [inline Libraries/LibJS/AST.h]: Personal opinion: I think these should be named similar to the following.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Assignment_Operators

File: Libraries/LibJS/Lexer.cpp [hunk 4]
```diff
@@ -205,8 +212,23 @@ Token Lexer::next()
     } else if (m_current_char == EOF) {
         token_type = TokenType::Eof;
     } else {
+        bool found_three_char_token = false;
+        if (m_position+1 < m_source.length()) {
+            char secondChar = m_source[m_position];
+            char thirdChar = m_source[m_position+1];
```
Comment (sunverwerth) [inline Libraries/LibJS/Lexer.cpp]: `secondChar` should be `second_char` etc...
I think I forgot to change these in the original PR.

File: Libraries/LibJS/Parser.cpp [hunk 5]
```diff
@@ -215,6 +253,53 @@ NonnullOwnPtr<VariableDeclaration> Parser::parse_variable_declaration()
...
+    default:
+        init = parse_statement();
+        break;
+    }
+
+    consume(TokenType::Semicolon);
+
+    OwnPtr<Expression> test = nullptr;
+    switch (m_current_token.type()) {
+    case TokenType::Semicolon:
```
Comment (sunverwerth) [inline Libraries/LibJS/Parser.cpp]: You can use `if (match(TokenType::Semicolon))` here instead

**Comment by sunverwerth** [comment]
@awesomekling  Ahh, you are too quick :rocket:

**Comment by awesomekling** [comment]
> @awesomekling Ahh, you are too quick

We'll just fix it incrementally! Great feedback btw :)

File: Libraries/LibJS/AST.h (hunk 3 above)
Comment (deoxxa) [reply Libraries/LibJS/AST.h]: Yep, I'd agree with that.

File: Libraries/LibJS/Parser.cpp (hunk 2 above)
Comment (sunverwerth) [reply Libraries/LibJS/Parser.cpp]: Oh, good point. I should refer to the spec more.

**Pull request merged**